File: solutions/go/interest-is-interesting/1/interest_is_interesting.go
package interest

func InterestRate(balance float64) float32 {
	switch {
    case balance < 0:
        return float32(3.213)
    case balance < 1000:
        return float32(0.5)
    case balance < 5000:
        return float32(1.621)
    default:
        return float32(2.475)
    }
}

func Interest(balance float64) float64 {
    rate := InterestRate(balance)
    return balance * float64(rate) / 100
}

func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
    years := 0
    current := balance
    for current < targetBalance {
        current += Interest(current)
        years++
    }
    return years
}

File: solutions/csharp/interest-is-interesting/1/InterestIsInteresting.cs
static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        switch (balance)
        {
            case < 0:
                return 3.213f;
            case < 1000:
                return 0.5f;
            case >= 1000 and < 5000:
                return 1.621f;
            default:
                return 2.475f;
        }
    }

    public static decimal Interest(decimal balance)
    {
        float rate = InterestRate(balance);
        return balance * (decimal)rate/100m;
    }

    public static decimal AnnualBalanceUpdate(decimal balance) => balance + Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int years = 0;
        decimal current = balance;
        while (current < targetBalance)
        {
            current += Interest(current);
            years++;
        }
        return years;
    }
}
